### ISSUE TITLE
spike: move from org admin to PSK

### ISSUE THREAD

**Issue by infinitewarp** [cloudigrade/cloudigrade#913]
## Summary

As a cloudigrade sysop, I want cloudigrade to keep working once the platform abandons the org admin flag and more fully embraces RBAC.

In the recent c.rh.c tech talk, it was mentioned that sources-api will soon abandon the org admin flag in favor of using RBAC. We need to figure out how this affects how we handle requests and if/when/how we need to also integrate with RBAC.

See also https://issues.redhat.com/browse/RHCLOUD-11483

> Currently sources is checking for if org admin then write permissions, and then if not org admin if the sourcesadmin role present then write permissions.
>
> Next we want to STOP checking for org admin and only check for the sources role. So this will change the behavior or current org admins, however if we add the sourcesadmin role to the "Default permission group" then customers can add or remove that as they see fit.

Questions we need to answer:

- Who is coordinating this effort to move from org admin to RBAC?
- When does org admin go away?
- Are all services related to us being pushed to use RBAC instead of org admin?
- Precisely what roles/etc. do we need to use from RBAC for restricting access?
  - For example, should we just use sourcesadmin role? Is there also a swatch-related admin role? Should we require both? Or something else?
- How exactly are we exposed to a user's roles? Do they come in the 3scale header? Do we have to call RBAC ourselves?

Expected output: answers to those questions and a new issue created to implement our change as needed.

Time box: 2 days

**Comment by abellotti** [comment]
- Who is coordinating this effort to move from org admin to RBAC?
  - No specifics yet, but most like Drew Bomhof and Jacob Lindgren.
  
- When does org admin go away?
  - Details haven't been fleshed out yet.

- Are all services related to us being pushed to use RBAC instead of org admin?
  - Sources is the last service that still allowed org_admin, all other services use Role/Permission as RBAC.

- Precisely what roles/etc. do we need to use from RBAC for restricting access?
   For example, should we just use sourcesadmin role? Is there also a swatch-related admin role? Should we require both? Or something else?
  - Details haven't been fleshed out yet, this will be discussed and communicated with all the teams before this major change goes in.

- How exactly are we exposed to a user's roles? Do they come in the 3scale header? Do we have to call RBAC ourselves?

X-RH-IDENTITY headers only contains the assigned entitlements to the user, but not roles or permissions. These have to be queried separately by calling the RBAC service. The X-RH-IDENTITY can be passed to the rbac service for authentication to query the related roles or permissions:

  -  External URL for RBAC for querying the roles:
            We get the roles from: https://cloud.redhat.com/api/rbac/v1/roles/?scope=principal
```
        data: [
          {   
            "name": "Approval User",
            ... 
          },  
          {   
            "name": "Compliance Administrator",
            ... 
          },  
          ... 
        ] 
```

We get the permissions from: https://cloud.redhat.com/api/rbac/v1/access/?application=

```
        data: [
          { 
            "permission": "inventory:*:*",
          },
          { 
            "permission": "advisor:*:*",
          },
          ...
        ]
```

Though from cloudigrade within the cluster, as per     https://docs.google.com/document/d/1xQqgxkZgIQTHpfvBpH_2F2Rpm_ox-LJcZDl5gxeW2vs/edit at the end of the document as well as the deployed services spreadsheet: https://docs.google.com/spreadsheets/d/188uz04yPb0Oe4rcq6QxLV1rHBjAJtp5UbUNZlPD2pwQ/edit#gid=1930857575 we can call the local rbac service and pass it the X-RH-IDENTITY from 3scale we have:

```
      http://rbac.rbac-prod.svc.cluster.local:8080
      http://rbac.rbac-stage.svc.cluster.local:8080
      http://rbac.rbac-qa.svc.cluster.local:8080
      http://rbac.rbac-ci.svc.cluster.local:8080
```

This is only available with 3scale X-RH-IDENTITY headers, sending in the X-RH-IDENTITY header from turnpike we give us a 40x error since the principal for internal SSO is different from 3scale.

For turnpike API calls, policy enforcement needed for internal endpoints is done via Turnpike policies directly, for which Turnpike checks the "associate" -> "role" -> [ ... ] in the X-RH-IDENTITY for policy enforcement.

API calls with Turnpike are authZ enforced per route so services don't need to enforce it: see: https://gitlab.cee.redhat.com/service/app-interface/-/blob/master/data/services/insights/turnpike/deploy.yml#L224-228
auth-Z per route is based on a user Role (i.e. from X-RH-IDENTITY associate->role[ ... ])

```
        i.e. turnpike X-RH-IDENTITY:
                "identity": {
                  "associate": {
                    "Role": [
                      "manage-account",
                      "view-profile",
                      "insights-services",
                      ...
                    ],
                    "email"...
                  }
                }
```

For service to service internal communication, there is no authZ as Turnpike only triggers the authZ check if we hit https://internal.cloud.redhat.com/api/* directory.  We cannot pass the Turnpike identity header to an internal api call, there is no authZ as it's only in turnpike.  For internal calls behind turnpike, best if we create the X-RH-IDENTITY using the X-RH-INTERNAL-FAKE-IDENTITY account_number and have sources do the rbac check. We can also do the same  for internal calls in cloudmeter by doing an RBAC check given the account_number from X-RH-INTERNAL-FAKE-IDENTITY.

Implementation wise, there is a shared common library for Ruby services (Sources, Topological Inventory, Approval, Catalog) provided here: https://github.com/RedHatInsights/insights-api-common-rails/tree/master/lib/insights/api/common/rbac. However, there does not seem to be a python shared library to access rbac, but services like koku does have a nice interface to it that we could model after: https://github.com/project-koku/koku/blob/master/koku/koku/rbac.py. We can create a Python shared library to do the whole lot and shared it between micro-services or just implement a subset as needed for now in cloudmeter.

**Comment by infinitewarp** [comment]
Thanks for the details notes, @abellotti!

Since it looks like we still have some important open questions about "who", "when", and "what", I'm moving this back to the backlog. Let's keep our eyes and ears open for any upcoming communication that may answer these questions. Otherwise, let's check back with people in a couple of weeks to ask if they have any answers yet.

**Comment by abellotti** [comment]
There is a shift in the direction of  https://issues.redhat.com/browse/RHCLOUD-11483, Sources is going for https://issues.redhat.com/browse/RHCLOUD-12839 as an implementation.
Dropping the RBAC Role/permission check and moving toward a service-to-service PSK approach.  Benefits of this approach include:
- a simpler implementation
- we'd eliminate the need to the rbac calls (faster)
- no need for the additional rbac access library/new code.
- specifying the account number and PSK as headers to Sources instead.

We'd need to have the PSKs made available to us in all our cloudigrade environments, ci/qa/stage/prod.

Sources PR for initial support is here: https://github.com/RedHatInsights/sources-api/pull/410

The new headers to sources would be:
- x-rh-sources-psk
- x-rh-sources-account-number

**Comment by infinitewarp** [comment]
Email sent today from Kathryn Dixon:

> TLDR
> Sources is removing the org admin check.
> 
> Longer Explanation
> 
> Working with Product Security and also the platform experience team, the sources app needed to change how we are integrated with RBAC. We are currently a snowflake implementation that checks if org admin -> can do anything or if sources administer role -> can do anything. We will be REMOVING the if org admin check to be compliant with product security concerns and to stop being the only app that still checks this (org admin part). 
> 
> The sources administer role will be added to the default role group to help limit customer pain. We will also document this in our docs, and add a pendo message to notify customers.
> 
> What needs to happen next
> 
> Cost mgmt and RHEL mgmt add ons apps must complete some changes. Full description is here in this document [0]. Please feel free to ask questions on implementations in it at anytime.
> 
> We know that the platform has many requirements that app teams must complete. So we are asking that these changes are completed by September 15, 2021. We estimated the effort required here is low, but we know that this is an additional task for teams. 
> 
> Please reach out with any questions or issues with the September 15, 2021 date.
> 
> Our tracking EPIC is here [1] and cost mgmt subtask [2] and subswatch subtask [3] 
> 
> [0] https://docs.google.com/document/d/11kM9FK72BiO-We0dVPpJna089mfHfOfG48KaQkKxmvg/edit?usp=sharing 
> [1] https://issues.redhat.com/browse/RHCLOUD-15285
> [2] https://issues.redhat.com/browse/RHCLOUD-15286
> [3] https://issues.redhat.com/browse/RHCLOUD-15287

**Comment by infinitewarp** [comment]
Given the new information above, I think we call this research spike done. We will open a new issue shortly to implement our changes.
